pkg/metrics/controller_test.go

```diff
@@ -284,3 +285,36 @@ func TestControllerGameServersNodeState(t *testing.T) {
 		{labels: []string{"false"}, val: int64(2)},
 	})
 }
+
+func TestCalcDuration(t *testing.T) {
+	m := agtesting.NewMocks()
+	c := NewController(
+		m.KubeClient,
+		m.AgonesClient,
+		m.KubeInformerFactory,
+		m.AgonesInformerFactory)
+	gs1 := gameServerWithFleetAndState("test-fleet", "")
+	gs1.ObjectMeta.CreationTimestamp = metav1.Now()
+	gs2 := gameServerWithFleetAndState("test-fleet", agonesv1.GameServerStateCreating)
+	gs2.ObjectMeta.CreationTimestamp = gs1.ObjectMeta.CreationTimestamp
+	duration, err := c.calcDuration(gs1, gs2)
+	assert.NoError(t, err, "Unable to caculate duration of a particular state")
+	assert.True(t, duration > 0., "Time diff should be calculated properly")
+	gs3 := gameServerWithFleetAndState("test-fleet", agonesv1.GameServerStateRequestReady)
+	gs3.ObjectMeta.CreationTimestamp = gs1.ObjectMeta.CreationTimestamp
+	duration, err = c.calcDuration(gs2, gs3)
+	assert.NoError(t, err, "Unable to caculate duration of a particular state")
+	assert.True(t, duration > 0., "Time diff should be calculated properly")
+
+	// gs1 state should already be deleted, error should be generated
+	// emulation of evicted key for gs1
+	duration, err = c.calcDuration(gs1, gs3)
+	assert.Error(t, err, "We should receive an error, metric should not be measured")
+	assert.True(t, duration == 0., "Time diff should be calculated properly")
```

Comment:
You could inject time.Now in the Controller.

now func() time.Time so that you can fake now and have a better expectations than >= 0